crates/ra_syntax/src/grammar/items/use_item.rs

```diff
@@ -7,25 +7,75 @@ pub(super) fn use_item(p: &mut Parser) {
     p.expect(SEMI);
 }
 
+/// Parse a use 'tree', such as `some::path` in `use some::path;`
+/// Note that this is called both by `use_item` and `use_tree_list`,
+/// so handles both `some::path::{inner::path}` and `inner::path` in
+/// `use some::path::{inner::path};`
 fn use_tree(p: &mut Parser) {
     let la = p.nth(1);
     let m = p.start();
     match (p.current(), la) {
+        // Finish the use_tree for cases of e.g.
+        // `use some::path::{self, *};` or `use *;`
+        // This does not handle cases such as `use some::path::*`
+        // N.B. in Rust 2015 `use *;` imports all from crate root
+        // however in Rust 2018 `use *;` errors: ('cannot glob-import all possible crates')
+        // TODO: Add this error (if not out of scope)
+
+        // test use_star
+        // use *;
+        // use ::*;
+        // use some::path::{*};
+        // use some::path::{::*};
         (STAR, _) => p.bump(),
         (COLONCOLON, STAR) => {
+            // Parse `use ::*;`, which imports all from the crate root in Rust 2015
+            // This is invalid inside a use_tree_list, (e.g. `use some::path::{::*}`)
+            // but still parses and errors later: ('crate root in paths can only be used in start position')
+            // TODO: Add this error (if not out of scope)
```

Comment:
We should add this error checking, but to the validator and not the parser:

https://github.com/rust-analyzer/rust-analyzer/blob/0b1c0ee225a075d25d700fcd1c007615556e2afb/crates/ra_syntax/src/validation/mod.rs

Reply:
Yeah, it was just an overall comment that it was something that naturally comes up there, but should be added somewhere else.